src/ssl.cpp

```diff
@@ -0,0 +1,219 @@
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/ssl.hpp"
+
+#if TORRENT_USE_SSL
+
+#ifdef TORRENT_USE_OPENSSL
+#include <openssl/x509v3.h> // for GENERAL_NAME
+#endif
+
+#ifdef TORRENT_USE_GNUTLS
+#include <gnutls/x509.h>
+#endif
+
+namespace libtorrent {
+namespace ssl {
+
+void set_trust_certificate(native_context_type nc, string_view pem, error_code &ec)
+{
+#if defined TORRENT_USE_OPENSSL
+	// create a new X.509 certificate store
+	X509_STORE* cert_store = X509_STORE_new();
+	if (!cert_store)
+	{
+		ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		return;
+	}
+
+	// wrap the PEM certificate in a BIO, for openssl to read
+	BIO* bp = BIO_new_mem_buf(
+			const_cast<char*>(pem.data()),
+			int(pem.size()));
+
+	// parse the certificate into OpenSSL's internal representation
+	X509* cert = PEM_read_bio_X509_AUX(bp, nullptr, nullptr, nullptr);
+	BIO_free(bp);
+
+	if (!cert)
+	{
+		X509_STORE_free(cert_store);
+		ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		return;
+	}
+
+	// add cert to cert_store
+	X509_STORE_add_cert(cert_store, cert);
+	X509_free(cert);
+
+	// and lastly, replace the default cert store with ours
+	SSL_CTX_set_cert_store(nc, cert_store);
+
+#elif defined TORRENT_USE_GNUTLS
+    gnutls_datum_t ca;
+    ca.data = reinterpret_cast<unsigned char*>(const_cast<char*>(pem.data()));
+    ca.size = unsigned(pem.size());
+
+	// Warning: returns the number of certificates processed or a negative error code on error
+	int ret = gnutls_certificate_set_x509_trust_mem(nc, &ca, GNUTLS_X509_FMT_PEM);
+	if(ret < 0)
+		ec = error_code(ret, error::get_ssl_category());
+#endif
+}
+
+void set_server_name_callback(context_handle_type c, server_name_callback_type cb, void* arg, error_code& ec)
+{
+#if defined TORRENT_USE_OPENSSL
+	TORRENT_UNUSED(ec);
+#if defined __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#endif
+	SSL_CTX_set_tlsext_servername_callback(c, cb);
+	SSL_CTX_set_tlsext_servername_arg(c, arg);
+#if defined __clang__
+#pragma clang diagnostic pop
+#endif
+
+#elif defined TORRENT_USE_GNUTLS
+	if(cb)
+		c->set_server_name_callback(
+				[cb, arg](stream_base& s, std::string const& name)
+				{
+					return cb(&s, name, arg);
+				}
+				, ec);
+	else
+		c->set_server_name_callback(nullptr);
+#endif
+}
+
+void set_host_name(stream_handle_type s, std::string const& name, error_code& ec)
+{
+#if defined TORRENT_USE_OPENSSL
+	TORRENT_UNUSED(ec);
+#if defined __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#endif
+	SSL_set_tlsext_host_name(s, name.c_str());
```

Comment:
I'm getting this warning:
```
../src/ssl.cpp:130:30: warning: cast from 'const char *' to 'char *' drops const qualifier [-Wcast-qual]
```

I think you also need:

```
#pragma clang diagnostic ignored "-Wcast-qual"
```

before this patch, all warnings are suppressed when using these openssl macros, by `#include "libtorrent/aux_/disable_warnings_push.hpp"`